### ISSUE TITLE
Scale IPv8 walk_interval

### ISSUE THREAD

**Issue by qstokkink** [Tribler/tribler#5384]
**tl;dr** You can gain additional download performance by scaling IPv8's `walk_interval`.

### Introduction
IPv8's inner loop works by scheduling activity at a certain interval (the `walk_interval`). This means that IPv8 will attempt to let every registered strategy update at this rate, e.g. the RandomWalk strategy will attempt to resolve a new peer every 0.5 seconds by default. We recently added a `health` REST endpoint to observe how well this `walk_interval` is met (in terms of additional imposed delay over the requested rate). If the main Python thread is busy doing other stuff (file I/O or CPU intensive activities) we see this reflected in the IPv8 health.

### Problem
If IPv8 is in bad health, it will continuously clog the main thread. Whereas this is not a problem when running IPv8 standalone (in which case it will simply be running at maximum speed), this does slow down external activities on the main thread as well -- like libtorrent in Tribler.

### Solution
By reflecting on the `IPv8` health (retrieved from the `health` REST endpoint) we can dynamically set the `walk_interval` of IPv8 to keep the main thread free for other stuff. However, the manner of scaling (i.e. what the minimum and maximum value should be and how you should get there) is an open problem.

**Comment by qstokkink** [comment]
As a quick showcase of how bad this can be, I used the same torrent with two different `walk_interval` settings. Note the difference in download speeds.

### Updating every 0.1 seconds
![afbeelding](https://user-images.githubusercontent.com/3630389/85003160-24bea880-b156-11ea-8e09-2351c295e4fd.png)

### Updating every 10.0 seconds
![afbeelding](https://user-images.githubusercontent.com/3630389/85003216-3738e200-b156-11ea-880e-f18e3606ecaf.png)

**Comment by ichorid** [comment]
The proper solution to this problem would be separating control traffic from bulk data transfer traffic into different endpoints/threads/processes.

**Comment by qstokkink** [comment]
> 
> 
> The proper solution to this problem would be separating control traffic from bulk data transfer traffic into different endpoints/threads/processes.

Whereas I agree that would certainly improve the QoS for Tribler and we need to do that at some point (in fact I already have an issue open in IPv8 for this): this issue is about a different problem -- one that is also much easier to solve. 

To keep the discussion focused: the issue here is about IPv8 (1) over-scheduling tasks on whatever event thread/process it is actually running in and (2) not detecting this and scaling back.

**Comment by ichorid** [comment]
Oh, I get it now. It should be easy to add timeouts or something like that.

**Comment by qstokkink** [comment]
Here is a prototype implementation for reference:

```patch
diff --git a/src/tribler-core/tribler_core/session.py b/src/tribler-core/tribler_core/session.py
index 681062458..376f2bf18 100644
--- a/src/tribler-core/tribler_core/session.py
+++ b/src/tribler-core/tribler_core/session.py
@@ -146,6 +146,36 @@ class Session(TaskManager):
         self.payout_manager = None
         self.mds = None  # Metadata Store
 
+        self.measuring_drift = None
+        self.scaling_reset_clock = 0
+
+    def auto_scale_ipv8(self):
+        import statistics
+        from ipv8.REST.health_endpoint import DriftMeasurementStrategy
+        default_walk_interval = 0.5
+        if self.measuring_drift is None:
+            # Enable
+            strategy = DriftMeasurementStrategy(self.ipv8.walk_interval)
+            with self.ipv8.overlay_lock:
+                self.ipv8.strategies.append((strategy, -1))
+            self.measuring_drift = strategy
+        else:
+            # Disable and scale
+            with self.ipv8.overlay_lock:
+                self.ipv8.strategies = [s for s in self.ipv8.strategies if not s[0] == self.measuring_drift]
+            history = [record[1] for record in self.measuring_drift.history]
+            median_time_taken = statistics.median(history) if history else default_walk_interval
+            if median_time_taken > self.ipv8.walk_interval/(len(self.ipv8.strategies) + 1.0):
+                self.ipv8.walk_interval = min(max(default_walk_interval, default_walk_interval + median_time_taken),
+                                              4 * default_walk_interval)
+                self._logger.info(f"Core choke detected: scaling IPv8 walk interval to {self.ipv8.walk_interval}.")
+            else:
+                self._logger.info("Everything seems fine: not scaling IPv8.")
+            self.scaling_reset_clock = (1 + self.scaling_reset_clock) % 20
+            if self.scaling_reset_clock == 0:
+                self.ipv8.walk_interval = max(default_walk_interval, self.ipv8.walk_interval/2.0)
+                self._logger.info(f"Optimistically scaling IPv8 walk interval down to {self.ipv8.walk_interval}.")
+
     def load_ipv8_overlays(self):
         if self.config.get_testnet():
             peer = Peer(self.trustchain_testnet_keypair)
@@ -444,6 +474,8 @@ class Session(TaskManager):
             if self.config.get_tunnel_community_enabled():
                 await self.tunnel_community.wait_for_socks_servers()
 
+            self.register_task("IPv8 auto-scaling", self.auto_scale_ipv8, interval=6.0)
+
         # Note that currently we should only start libtorrent after the SOCKS5 servers have been started
         if self.config.get_libtorrent_enabled():
             self.readable_status = STATE_START_LIBTORRENT

```

**Comment by ichorid** [comment]
Maybe it would be easy to just add a timeout to every IPv8 walk call, so it just silently skips the beat if it could not do it in time?

**Comment by qstokkink** [comment]
> Maybe it would be easy to just add a timeout to every IPv8 walk call, so it just silently skips the beat if it could not do it in time?

Periodic choking of the core is no problem. It just resumes normally after the spike, it wont be permanently choked.

Specifically, this issue is about **always** choking. If the IPv8 core can **never** schedule events on time, there is a problem.

**Comment by qstokkink** [comment]
I have the code ready for this (https://github.com/qstokkink/tribler/tree/add_ipv8_scaling) and we can make a PR whenever we want this.
